eval.go

```diff
@@ -587,7 +577,29 @@ func (d *indexData) gatherBranches(docID uint32, mt matchTree, known map[matchTr
 			id <<= 1
 			mask >>= 1
 		}
+
+		return branches
 	}
+
+	foundBranchQuery := false
+	var branches []string
+	visitMatches(mt, known, func(mt matchTree) {
+		bq, ok := mt.(*branchQueryMatchTree)
+		if !ok {
+			return
+		}
+		foundBranchQuery = true
+		mask := bq.masks[repoIdx] & bq.fileMasks[docID]
+		branches = append(branches, gather(mask)...)
+	})
+
+	// If the query doesn't contain a "branch:" filter, we return all branches for
+	// docID.
+	if !foundBranchQuery {
+		mask := d.fileBranchMasks[docID]
+		branches = gather(mask)
+	}
```

Comment:
A suggestion whlie you are touching this code. I find it kind of weird how we append to branches. Why don't we just `|` the masks found by `branchQueryMatchTree`. Then if the final mask is `0`, we treat it like foundBranchQuery is false. Alternatively if it is possible for a mask in a branchQueryMatchTree to be 0, then we keep the foundBranchQuery boolean.

This has the benefit of always outputting branches in the order they appear in the fileBranchMasks. Additionally we could even do micro optimizations like using popcount to know the size of branches :P